wpilib/src/i2c.rs

```diff
@@ -0,0 +1,187 @@
+use wpilib_sys::*;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[repr(i32)]
+pub enum Port {
+    Onboard = HAL_I2CPort::HAL_I2C_kOnboard,
+    MXP = HAL_I2CPort::HAL_I2C_kMXP,
+}
+
+pub struct I2C {
+    port: Port,
+    device_address: i32,
+}
+
+impl I2C {
+    /// Constructs a new I2C
+    ///
+    /// `port` is the I2C port to which the device is connected, and `device_address` is the address of the device on the bus
+    pub fn new(port: Port, device_address: i32) -> HalResult<Self> {
+        hal_call!(HAL_InitializeI2C(port as HAL_I2CPort::Type))?;
+        usage::report(usage::resource_types::I2C, 0);
+        Ok(I2C {
+            port,
+            device_address,
+        })
+    }
+
+    /// Generic transaction.
+    ///
+    /// This is a lower-level interface to the I2C hardware giving you more control
+    /// over each transaction.
+    ///
+    /// This function will send all the bytes in `data_to_send` and will read data into `data_received`. Callers should make sure these buffers are sized appropriately
+    ///
+    /// Returns true if the transfer was aborted, false if the transfer was completed successfully
+    pub fn transaction(&self, data_to_send: &[u8], data_received: &mut [u8]) -> bool {
+        let status = unsafe {
+            HAL_TransactionI2C(
+                self.port as HAL_I2CPort::Type,
+                self.device_address,
+                data_to_send.as_ptr(),
+                data_to_send.len() as i32,
+                data_received.as_mut_ptr(),
+                data_received.len() as i32,
+            )
+        };
+
+        status < 0
+    }
+
+    /// Attempt to address a device on the I2C bus.
+    ///
+    /// This allows you to figure out if there is a device on the I2C bus that
+    /// responds to the address specified in the constructor.
+    ///
+    /// Returns false if the transaction was successful, true if it was aborted
+    pub fn address_only(&self) -> bool {
+        self.transaction(&[], &mut [])
+    }
+
+    /// Execute a write transaction with the device.
+    ///
+    /// Write a single byte to a register on a device and wait until the
+    ///   transaction is complete.
+    ///
+    /// Returns false if the transaction was successful, true if it was aborted
+    pub fn write(&self, register_address: i32, data: u8) -> bool {
+        let mut buf = [0u8; 2];
+        buf[0] = register_address as u8; // TODO: Is this valid?
```

Comment:
I2C registers can't be larger than a u8, so this is technically correct. Should take a u8 for the address though.